File: java/src/main/java/com/experian/dvs/client/server/address/search/RestApiAddressSearchRequest.java
package com.experian.dvs.client.server.address.search;

import com.experian.dvs.client.address.*;
import com.experian.dvs.client.server.address.Address;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.ArrayList;
import java.util.List;

public class RestApiAddressSearchRequest {

    @JsonProperty("country_iso")
    private String countryIso;

    @JsonProperty("datasets")
    private List<String> datasets;

    @JsonProperty("max_suggestions")
    private Integer maxSuggestions;

    @JsonProperty("components")
    private Address address;

    @JsonProperty("location")
    private String location;

    @JsonProperty("options")
    private List<RestApiAdditionalOption> options;

    public static RestApiAddressSearchRequest using(final AddressConfiguration configuration) {
        RestApiAddressSearchRequest searchRequest = new RestApiAddressSearchRequest();

        //Country
        if (configuration.getCountry() != null) {
            searchRequest.setCountryIso(configuration.getCountry().getIso3Code());
        }
        //Datasets
        if (!configuration.getDatasets().isEmpty()) {
            searchRequest.setDatasets(configuration.getDatasets().stream().map(Dataset::getDatasetCode).toList());
        }
        //Max suggestions
        if (configuration.getMaxSuggestions() != AddressConfiguration.DEFAULT_MAX_SUGGESTIONS) {
            searchRequest.setMaxSuggestions(configuration.getMaxSuggestions());
        }
        //Location
        if (!configuration.getLocation().isEmpty()) {
            searchRequest.setLocation(configuration.getLocation());
        }

        if (configuration.getFlattenResults()) {
            searchRequest.getOptions().add(new RestApiAdditionalOption("flatten", Boolean.TRUE.toString()));
        }
        if (configuration.getSearchIntensity() != null) {
            searchRequest.getOptions().add(new RestApiAdditionalOption("intensity", configuration.getSearchIntensity().getName()));
        }
        if (configuration.getPromptSet() != null) {
            searchRequest.getOptions().add(new RestApiAdditionalOption("prompt_set", configuration.getPromptSet().getName()));
        }

        //todo preferred language and preferred script


        return searchRequest;
    }

    public String getCountryIso() {
        return countryIso;
    }

    public void setCountryIso(String countryIso) {
        this.countryIso = countryIso;
    }

    public List<String> getDatasets() {
        return datasets;
    }

    public void setDatasets(List<String> datasets) {
        this.datasets = datasets;
    }

    public Address getAddress() {
        return address;
    }

    public void setAddress(final Address address) {
        this.address = address;
    }

    public List<RestApiAdditionalOption> getOptions() {
        return options;
    }

    public void setOptions(List<RestApiAdditionalOption> options) {
        this.options = options;
    }

        public void addOption(final String name, final String value) {
            if (this.options == null) {
                this.options = new ArrayList<>();
            }
            this.options.add(new RestApiAdditionalOption(name, value));
        }

    public Integer getMaxSuggestions() {
        return maxSuggestions;
    }

    public void setMaxSuggestions(Integer maxSuggestions) {
        this.maxSuggestions = maxSuggestions;
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
    }
}

File: csharp/DVSClient/DVSClient/server/address/search/RestApiAddressSearchRequest.cs
using Newtonsoft.Json;
using DVSClient.Address;

namespace DVSClient.Server.Address.Search
{
    public class RestApiAddressSearchRequest
    {
        [JsonProperty("country_iso")]
        public string? CountryIso { get; set; }

        [JsonProperty("datasets")]
        public IEnumerable<string>? Datasets { get; set; }

        [JsonProperty("max_suggestions")]
        public int? MaxSuggestions { get; set; }

        [JsonProperty("components")]
        public Address? Address { get; set; }

        [JsonProperty("location")]
        public string? Location { get; set; }

        [JsonProperty("options")]
        public IList<RestApiAdditionalOption>? Options { get; set; }

        public static RestApiAddressSearchRequest Using(AddressConfiguration configuration)
        {
            var searchRequest = new RestApiAddressSearchRequest();

            // Country
            if (configuration.GetCountry() != null)
            {
                searchRequest.CountryIso = configuration.GetCountry()?.Iso3Code;
            }

            // Datasets
            if (configuration.Datasets.Any())
            {
                searchRequest.Datasets = configuration.Datasets.ToList().ConvertAll(dataset => dataset.DatasetCode);
            }

            // Max suggestions
            if (configuration.MaxSuggestions != AddressConfiguration.DefaultMaxSuggestions)
            {
                searchRequest.MaxSuggestions = configuration.MaxSuggestions;
            }

            // Location
            if (!string.IsNullOrEmpty(configuration.Location))
            {
                searchRequest.Location = configuration.Location;
            }

            if (configuration.FlattenResults)
            {
                searchRequest.AddOption("flatten", true.ToString());
            }

            if (configuration.SearchIntensity.HasValue)
            {
                searchRequest.AddOption("intensity", configuration.SearchIntensity.Value.ToString());
            }

            if (configuration.PromptSet.HasValue)
            {
                searchRequest.AddOption("prompt_set", configuration.PromptSet.Value.ToString());
            }

            // TODO: Add preferred language and preferred script

            return searchRequest;
        }

        public void AddOption(string name, string value)
        {
            if (Options == null)
            {
                Options = new List<RestApiAdditionalOption>();
            }

            Options.Add(new RestApiAdditionalOption(name, value));
        }
    }
}